### PR TITLE
Add port for Linux on LoongArch

### PR DESCRIPTION
**Pull request by tangyouling**
This commit adds support for Linux on LoongArch. Only 64-bit is supported
at the moment.

The changes were tested on a LoongArch architecture Loongson-3A5000
processor (LoongArch is a new RISC ISA, which is a bit like MIPS or
RISC-V.).
```
$ uname -a
Linux tang-pc 4.19.190+ #177 SMP Mon Oct 25 09:06:23 CST 2021 loongarch64 
loongarch64 loongarch64 GNU/Linux
```

```
After applying the patch, the test results is as follows:
$ make check
 ...
 FAIL: Ltest-cxx-exceptions
 PASS: run-check-namespace
 PASS: run-ptrace-mapper
 PASS: run-ptrace-misc
 PASS: run-coredump-unwind
 PASS: run-coredump-unwind-mdi
 ============================================================================
 Testsuite summary for libunwind 1.6-rc1
 ============================================================================
 # TOTAL: 36
 # PASS:  35
 # SKIP:  0
 # XFAIL: 0
 # FAIL:  1
 ...
```

When the commit "f1cee65e7594 (Link Ltest-cxx-exceptions with libunwind)" is
reverted, all test cases can be passed.

Signed-off-by: Tiezhu Yang <yangtiezhu@loongson.cn>
Signed-off-by: Qing Zhang <zhangqing@loongson.cn>
Signed-off-by: Jinyang He <hejinyang@loongson.cn>
Signed-off-by: Youling Tang <tangyouling@loongson.cn>

### CODE REVIEW

File: src/loongarch64/Ginit.c [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+  else
+    {
+      *val = (unw_word_t) *(unw_word_t *) (intptr_t) addr;
+      Debug (12, "%s -> %llx\n", unw_regname (reg), (long long) *val);
+    }
+  return 0;
+
+ badreg:
+  Debug (1, "bad register number %u\n", reg);
+  return -UNW_EBADREG;;
```
Comment (yetist) [inline src/loongarch64/Ginit.c:366]: Too many semicolons.

**Review by yetist — COMMENTED** [review]
I used the following command line to compile:
```
autoreconf -ifv
./configure --prefix=/usr
make -j8
```

But I got an error when linking:
```
libtool: link: gcc -shared  -fPIC -DPIC  .libs/os-linux.o .libs/dl-iterate-phdr.o mi/.libs/init.o mi/.libs/flush_cache.o mi/.libs/mempool.o mi/.libs/strerror.o loongarch64/.libs/is_fpreg.o loongarch64/.libs/regname.o mi/.libs/_ReadULEB.o mi/.libs/_ReadSLEB.o mi/.libs/backtrace.o mi/.libs/dyn-cancel.o mi/.libs/dyn-info-list.o mi/.libs/dyn-register.o mi/.libs/Ldyn-extract.o mi/.libs/Lfind_dynamic_proc_info.o mi/.libs/Lget_accessors.o mi/.libs/Lget_proc_info_by_ip.o mi/.libs/Lget_proc_name.o mi/.libs/Lput_dynamic_unwind_info.o mi/.libs/Ldestroy_addr_space.o mi/.libs/Lget_reg.o mi/.libs/Lset_reg.o mi/.libs/Lget_fpreg.o mi/.libs/Lset_fpreg.o mi/.libs/Lset_caching_policy.o mi/.libs/Lset_cache_size.o unwind/.libs/Backtrace.o unwind/.libs/DeleteException.o unwind/.libs/FindEnclosingFunction.o unwind/.libs/ForcedUnwind.o unwind/.libs/GetBSP.o unwind/.libs/GetCFA.o unwind/.libs/GetDataRelBase.o unwind/.libs/GetGR.o unwind/.libs/GetIP.o unwind/.libs/GetLanguageSpecificData.o unwind/.libs/GetRegionStart.o unwind/.libs/GetTextRelBase.o unwind/.libs/RaiseException.o unwind/.libs/Resume.o unwind/.libs/Resume_or_Rethrow.o unwind/.libs/SetGR.o unwind/.libs/SetIP.o unwind/.libs/GetIPInfo.o loongarch64/.libs/getcontext.o loongarch64/.libs/Lapply_reg_state.o loongarch64/.libs/Lreg_states_iterate.o loongarch64/.libs/Lcreate_addr_space.o loongarch64/.libs/Lget_proc_info.o loongarch64/.libs/Lget_save_loc.o loongarch64/.libs/Lglobal.o loongarch64/.libs/Linit.o loongarch64/.libs/Linit_local.o loongarch64/.libs/Linit_remote.o loongarch64/.libs/Lis_signal_frame.o loongarch64/.libs/Lregs.o loongarch64/.libs/Lresume.o loongarch64/.libs/Lstep.o  -Wl,--whole-archive ./.libs/libunwind-dwarf-local.a ./.libs/libunwind-elfxx.a -Wl,--no-whole-archive  -lc -lgcc_s -llzma -lz  -g -O2 -nostartfiles -nostdlib   -Wl,-soname -Wl,libunwind.so.8 -o .libs/libunwind.so.8.0.1
/usr/bin/ld: loongarch64/.libs/Linit.o:(.bss+0x0): multiple definition of `_U_dyn_info_list'; mi/.libs/dyn-info-list.o:(.bss+0x0): first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:3683: libunwind.la] Error 1
make[2]: Leaving directory '/home/yetist/checkout/04-la64/libunwind/src'
make[1]: *** [Makefile:2448: all] Error 2
make[1]: Leaving directory '/home/yetist/checkout/04-la64/libunwind/src'
make: *** [Makefile:615: all-recursive] Error 1
```

**Comment by tangyouling** [comment]
> I used the following command line to compile:
> 
> ```
> autoreconf -ifv
> ./configure --prefix=/usr
> make -j8
> ```
> 
> But I got an error when linking:
> 
> ```
> libtool: link: gcc -shared  -fPIC -DPIC  .libs/os-linux.o .libs/dl-iterate-phdr.o mi/.libs/init.o mi/.libs/flush_cache.o mi/.libs/mempool.o mi/.libs/strerror.o loongarch64/.libs/is_fpreg.o loongarch64/.libs/regname.o mi/.libs/_ReadULEB.o mi/.libs/_ReadSLEB.o mi/.libs/backtrace.o mi/.libs/dyn-cancel.o mi/.libs/dyn-info-list.o mi/.libs/dyn-register.o mi/.libs/Ldyn-extract.o mi/.libs/Lfind_dynamic_proc_info.o mi/.libs/Lget_accessors.o mi/.libs/Lget_proc_info_by_ip.o mi/.libs/Lget_proc_name.o mi/.libs/Lput_dynamic_unwind_info.o mi/.libs/Ldestroy_addr_space.o mi/.libs/Lget_reg.o mi/.libs/Lset_reg.o mi/.libs/Lget_fpreg.o mi/.libs/Lset_fpreg.o mi/.libs/Lset_caching_policy.o mi/.libs/Lset_cache_size.o unwind/.libs/Backtrace.o unwind/.libs/DeleteException.o unwind/.libs/FindEnclosingFunction.o unwind/.libs/ForcedUnwind.o unwind/.libs/GetBSP.o unwind/.libs/GetCFA.o unwind/.libs/GetDataRelBase.o unwind/.libs/GetGR.o unwind/.libs/GetIP.o unwind/.libs/GetLanguageSpecificData.o unwind/.libs/GetRegionStart.o unwind/.libs/GetTextRelBase.o unwind/.libs/RaiseException.o unwind/.libs/Resume.o unwind/.libs/Resume_or_Rethrow.o unwind/.libs/SetGR.o unwind/.libs/SetIP.o unwind/.libs/GetIPInfo.o loongarch64/.libs/getcontext.o loongarch64/.libs/Lapply_reg_state.o loongarch64/.libs/Lreg_states_iterate.o loongarch64/.libs/Lcreate_addr_space.o loongarch64/.libs/Lget_proc_info.o loongarch64/.libs/Lget_save_loc.o loongarch64/.libs/Lglobal.o loongarch64/.libs/Linit.o loongarch64/.libs/Linit_local.o loongarch64/.libs/Linit_remote.o loongarch64/.libs/Lis_signal_frame.o loongarch64/.libs/Lregs.o loongarch64/.libs/Lresume.o loongarch64/.libs/Lstep.o  -Wl,--whole-archive ./.libs/libunwind-dwarf-local.a ./.libs/libunwind-elfxx.a -Wl,--no-whole-archive  -lc -lgcc_s -llzma -lz  -g -O2 -nostartfiles -nostdlib   -Wl,-soname -Wl,libunwind.so.8 -o .libs/libunwind.so.8.0.1
> /usr/bin/ld: loongarch64/.libs/Linit.o:(.bss+0x0): multiple definition of `_U_dyn_info_list'; mi/.libs/dyn-info-list.o:(.bss+0x0): first defined here
> collect2: error: ld returned 1 exit status
> make[2]: *** [Makefile:3683: libunwind.la] Error 1
> make[2]: Leaving directory '/home/yetist/checkout/04-la64/libunwind/src'
> make[1]: *** [Makefile:2448: all] Error 2
> make[1]: Leaving directory '/home/yetist/checkout/04-la64/libunwind/src'
> make: *** [Makefile:615: all-recursive] Error 1
> ```

As kasperk81 said, applying the following patch should solve the link problem

`diff --git a/src/loongarch64/Ginit.c b/src/loongarch64/Ginit.c
index 674e5b06..d2f308c7 100644
--- a/src/loongarch64/Ginit.c
+++ b/src/loongarch64/Ginit.c
@@ -63,8 +63,6 @@ tdep_uc_addr (ucontext_t *uc, int reg)
 
 # endif /* UNW_LOCAL_ONLY */
 
-HIDDEN unw_dyn_info_list_t _U_dyn_info_list;
-
 static void
 put_unwind_info (unw_addr_space_t as, unw_proc_info_t *proc_info, void *arg)
 {
@@ -75,7 +73,13 @@ static int
 get_dyn_info_list_addr (unw_addr_space_t as, unw_word_t *dyn_info_list_addr,
                         void *arg)
 {
-  *dyn_info_list_addr = (unw_word_t) (intptr_t) &_U_dyn_info_list;
+#ifndef UNW_LOCAL_ONLY
+# pragma weak _U_dyn_info_list_addr
+  if (!_U_dyn_info_list_addr)
+    return -UNW_ENOINFO;
+#endif
+  // Access the `_U_dyn_info_list` from `LOCAL_ONLY` library, i.e. libunwind.so.
+  *dyn_info_list_addr = _U_dyn_info_list_addr ();
   return 0;
 }`

**Comment by xen0n** [comment]
@tangyouling Please fix your Markdown formatting; the code block are garbled and very hard to read for others.

It seems you used the single-backtick quote form (inline code) for code blocks, which is inappropriate as it doesn't handle newlines; you should use [the fenced code block](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) instead. Refer to [the GFM help doc](https://docs.github.com/en/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) for additional details.

**Comment by kasperk81** [comment]
> When the commit "f1cee65e7594 (Link Ltest-cxx-exceptions with libunwind)" is

f1cee65e7594 adds the test in make file. Reverting that means disabling the only failing test. :)

Did you check `tests/test-suite.log` what is the error? You can try to debug it using gdb as well.

**Comment by kasperk81** [comment]
ok. this test was passing on risicv64 so I was wondering if there is something in this implementation which can be fixed to pass this test. but if you have checked `tests/test-suite.log` and you know loongarch by design doesn't support it, then feel free to disable it in configure.ac.

**Comment by tangyouling** [comment]
@kasperk81 
> ok. this test was passing on risicv64 so I was wondering if there is something in this implementation which can be fixed to pass this test. but if you have checked `tests/test-suite.log` and you know loongarch by design doesn't support it, then feel free to disable it in configure.ac.

Because I haven't found a way to fix it for the time being, and I don't know whether the loongarch design supports it. So I temporarily disable it in `configure.ac `

**Comment by djwatson** [comment]
Is this ready for commit?

**Comment by tangyouling** [comment]
> Is this ready for commit?

@djwatson 
Yes, I have resubmitted based on the latest commit
Sorry, because of a mistake in my operation, I resubmit a new Pull request. The link is as follows:https://github.com/libunwind/libunwind/pull/316

**Comment by djwatson** [comment]
Thanks!

**Pull request closed without merging**